<h2 style="color:red" align="center">Implement Neural Network (or Logistic Regression) From Scratch</h2>

<h4 style="color:blue">Predicting if a person will have diabetes or not depending upon 8 features using logistic regression</h4>

<h4 style="color:blue">Use the below link of the kaggle dataset for this project </h4>
<h5 style="color:blue"> https://www.kaggle.com/datasets/mathchi/diabetes-data-set </h5>

In [6]:
import numpy as np
import tensorflow as tf
from tensorflow import keras
import pandas as pd
from matplotlib import pyplot as plt
%matplotlib inline

In [7]:
df = pd.read_csv("./diabetes.csv")
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

Out of all the features the most important ones which affects the outcome the most are 'Pregnancies', 'BloodPressure', 'SkinThickness', 'BMI', ' DiabetesPedigreeFunction', and 'Age'. So dropping two columns "Glucose" and "Insulin"

In [8]:
dfn = df.drop(['Glucose', 'Insulin'], axis=1)

In [9]:
dfn.columns

Index(['Pregnancies', 'BloodPressure', 'SkinThickness', 'BMI',
       'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

In [10]:
dfn.shape

(768, 7)

In [11]:
dfn.describe()

Pregnancies  BloodPressure  SkinThickness         BMI  \
count   768.000000     768.000000     768.000000  768.000000   
mean      3.845052      69.105469      20.536458   31.992578   
std       3.369578      19.355807      15.952218    7.884160   
min       0.000000       0.000000       0.000000    0.000000   
25%       1.000000      62.000000       0.000000   27.300000   
50%       3.000000      72.000000      23.000000   32.000000   
75%       6.000000      80.000000      32.000000   36.600000   
max      17.000000     122.000000      99.000000   67.100000   

       DiabetesPedigreeFunction         Age     Outcome  
count                768.000000  768.000000  768.000000  
mean                   0.471876   33.240885    0.348958  
std                    0.331329   11.760232    0.476951  
min                    0.078000   21.000000    0.000000  
25%                    0.243750   24.000000    0.000000  
50%                    0.372500   29.000000    0.000000  
75%                    0.626250   41.000000    1.000000  
max                    2.420000   81.000000    1.000000

**Split train and test data**

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(dfn[['Pregnancies', 'BloodPressure', 'SkinThickness', 'BMI',
       'DiabetesPedigreeFunction', 'Age']],dfn.Outcome, test_size=0.2, random_state=25)

**Preprocessing: Scale the data so that all features are in the same range**

In [13]:
X_train_scaled = X_train.copy()
X_train_scaled['Pregnancies'] = X_train_scaled['Pregnancies'] / 17
X_train_scaled['BloodPressure'] = X_train_scaled['BloodPressure'] / 122
X_train_scaled['SkinThickness'] = X_train_scaled['SkinThickness'] / 99
X_train_scaled['BMI'] = X_train_scaled['BMI'] / 67.1
X_train_scaled['DiabetesPedigreeFunction'] = X_train_scaled['DiabetesPedigreeFunction'] / 2.42
X_train_scaled['Age'] = X_train_scaled['Age'] / 81

In [14]:
X_train_scaled.head(2)

Pregnancies  BloodPressure  SkinThickness       BMI  \
118     0.235294       0.491803       0.232323  0.420268   
336     0.000000       0.000000       0.000000  0.503726   

     DiabetesPedigreeFunction       Age  
118                  0.183058  0.271605  
336                  0.385124  0.543210

In [15]:
X_test_scaled = X_test.copy()
X_test_scaled['Pregnancies'] = X_test_scaled['Pregnancies'] / 17
X_test_scaled['BloodPressure'] = X_test_scaled['BloodPressure'] / 122
X_test_scaled['SkinThickness'] = X_test_scaled['SkinThickness'] / 99
X_test_scaled['BMI'] = X_test_scaled['BMI'] / 67.1
X_test_scaled['DiabetesPedigreeFunction'] = X_test_scaled['DiabetesPedigreeFunction'] / 2.42
X_test_scaled['Age'] = X_test_scaled['Age'] / 81

**Model Building using Keras first and then we will build our neural network from scratch**

In [16]:
model = keras.Sequential([
    keras.layers.Dense(1, input_shape=(6,), activation='sigmoid', kernel_initializer='ones', bias_initializer='zeros')
])

model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['accuracy'])

model.fit(X_train_scaled, y_train, epochs=5000)

Streaming output truncated to the last 5000 lines.
20/20 [==============================] - 0s 4ms/step - loss: 0.5513 - accuracy: 0.6938
Epoch 2502/5000
20/20 [==============================] - 0s 3ms/step - loss: 0.5513 - accuracy: 0.6922
Epoch 2503/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5512 - accuracy: 0.6922
Epoch 2504/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5512 - accuracy: 0.6922
Epoch 2505/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5514 - accuracy: 0.6889
Epoch 2506/5000
20/20 [==============================] - 0s 3ms/step - loss: 0.5512 - accuracy: 0.6906
Epoch 2507/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5512 - accuracy: 0.6922
Epoch 2508/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5512 - accuracy: 0.6922
Epoch 2509/5000
20/20 [==============================] - 0s 4ms/step - loss: 0.5512 - accuracy: 0.6922
Epoch 2510/5000
20/20 [===============

**Model Evaluation on test data**

In [17]:
model.evaluate(X_test_scaled,y_test)

5/5 [==============================] - 0s 5ms/step - loss: 0.5651 - accuracy: 0.7208


[0.5651317238807678, 0.7207792401313782]

In [18]:
model.predict(X_test_scaled)[:5]

5/5 [==============================] - 0s 5ms/step


array([[0.66839594],
       [0.78786856],
       [0.71803695],
       [0.6174387 ],
       [0.2994859 ]], dtype=float32)

In [19]:
y_test[:5]

459    0
39     1
344    0
84     1
700    0
Name: Outcome, dtype: int64

 get the value of weights and bias from the model

In [20]:
coef, intercept = model.get_weights()
coef, intercept

(array([[ 1.8236613 ],
        [-1.599353  ],
        [ 0.06977581],
        [ 6.785558  ],
        [ 2.07437   ],
        [ 2.6513433 ]], dtype=float32),
 array([-4.982498], dtype=float32))

This means w1=1.8236613, w2=-1.599353, w3=0.06977581, w4 = 6.785558, w5=2.07437, w6=2.6513433,  bias =-4.982498

In [21]:
def sigmoid(x):
        import math
        return 1 / (1 + math.exp(-x))
sigmoid(18)

0.9999999847700205

In [23]:
X_test_scaled[:5]

Pregnancies  BloodPressure  SkinThickness       BMI  \
459     0.529412       0.606557       0.333333  0.385991   
39      0.235294       0.590164       0.474747  0.552906   
344     0.470588       0.590164       0.000000  0.548435   
84      0.294118       0.885246       0.000000  0.727273   
700     0.117647       0.622951       0.272727  0.535022   

     DiabetesPedigreeFunction       Age  
459                  0.190083  1.000000  
39                   0.574380  0.691358  
344                  0.200413  0.703704  
84                   0.093802  0.456790  
700                  0.199587  0.320988

Instead of model.predict, write our own prediction function that uses w1, w2,w3, w4, w5, w6 and bias

In [24]:
def prediction_function(Pregnancies, BloodPressure, SkinThickness, BMI,
       DiabetesPedigreeFunction, Age):
    weighted_sum = coef[0]*Pregnancies + coef[1]*BloodPressure + coef[2]*SkinThickness + coef[3]*BMI + coef[4]*DiabetesPedigreeFunction + coef[5]*Age+ intercept
    return sigmoid(weighted_sum)

prediction_function(0.529412,	0.606557,	0.333333,	0.385991,	0.190083,	1.000000)

0.6683961785981375

Now we start implementing gradient descent in plain python. Again the goal is to come up with same w1, w2 and bias that keras model calculated. We want to show how keras/tensorflow would have computed these values internally using gradient descent
First write couple of helper routines such as sigmoid and log_loss

In [27]:
def sigmoid_numpy(X):
   return 1/(1+np.exp(-X))


In [28]:
def log_loss(y_true, y_predicted):
    epsilon = 1e-15
    y_predicted_new = [max(i,epsilon) for i in y_predicted]
    y_predicted_new = [min(i,1-epsilon) for i in y_predicted_new]
    y_predicted_new = np.array(y_predicted_new)
    return -np.mean(y_true*np.log(y_predicted_new)+(1-y_true)*np.log(1-y_predicted_new))

In [29]:
def gradient_descent(Pregnancies, BloodPressure, SkinThickness, BMI,
       DiabetesPedigreeFunction, Age, y_true, epochs, loss_thresold):
    w1 = w2 = w3 = w4 = w5 = w6 = 1
    bias = 0
    rate = 0.5
    n = len(Pregnancies)
    for i in range(epochs):
        weighted_sum = w1 * Pregnancies + w2 * BloodPressure + w3 * SkinThickness + w4 * BMI + w5 * DiabetesPedigreeFunction + w6 * Age +bias
        y_predicted = sigmoid_numpy(weighted_sum)
        loss = log_loss(y_true, y_predicted)

        w1d = (1/n)*np.dot(np.transpose(Pregnancies),(y_predicted-y_true))
        w2d = (1/n)*np.dot(np.transpose(BloodPressure),(y_predicted-y_true))
        w3d = (1/n)*np.dot(np.transpose(SkinThickness),(y_predicted-y_true))
        w4d = (1/n)*np.dot(np.transpose(BMI),(y_predicted-y_true))
        w5d = (1/n)*np.dot(np.transpose(DiabetesPedigreeFunction),(y_predicted-y_true))
        w6d = (1/n)*np.dot(np.transpose(Age),(y_predicted-y_true))

        bias_d = np.mean(y_predicted - y_true)
        w1 = w1 - rate * w1d
        w2 = w2 - rate * w2d
        w3 = w3 - rate * w3d
        w4 = w4 - rate * w4d
        w5 = w5 - rate * w5d
        w6 = w6 - rate * w6d
        bias = bias - rate * bias_d

        print (f'Epoch:{i}, w1:{w1}, w2:{w2}, w3:{w3}, w4:{w4}, w5:{w5}, w6:{w6}, bias:{bias}, loss:{loss}')

        if loss<=loss_thresold:
            break

    return w1, w2, w3, w4, w5, w6, bias

In [33]:
#loss_threshold is the threshold value of the loss at which we want to break the gradient descent
gradient_descent(X_train_scaled['Pregnancies'],X_train_scaled['BloodPressure'],X_train_scaled['SkinThickness'],X_train_scaled['BMI'],X_train_scaled['DiabetesPedigreeFunction'],X_train_scaled['Age'], y_train,3000, 0.558)

Epoch:0, w1:0.9484038037549335, w2:0.8482495061987809, w3:0.9464823618189773, w4:0.8811597148577026, w5:0.9526961436461479, w6:0.8966941323397016, bias:-0.26403054991935676, loss:1.4049969349785973
Epoch:1, w1:0.9024086995022679, w2:0.7122443605376363, w3:0.8983818658009682, w4:0.7754970806347241, w5:0.9105551976058015, w6:0.8046070238345744, bias:-0.4997355360184146, loss:1.167063221357098
Epoch:2, w1:0.8633218954716789, w2:0.5952209414050482, w3:0.8568989354648177, w4:0.68565850265964, w5:0.8746586822954513, w6:0.7260515442138235, bias:-0.7017713662980494, loss:0.9809885839064499
Epoch:3, w1:0.8317409286608748, w2:0.4983351500703417, w3:0.8225332043859629, w4:0.6125672888688332, w5:0.8454266384053672, w6:0.6618848383616367, bias:-0.8684809366915623, loss:0.8469962959635866
Epoch:4, w1:0.8073989579900275, w2:0.42044550913021894, w3:0.7949680890928095, w4:0.5552540289498824, w5:0.8225277894065064, w6:0.6113329650497608, bias:-1.0021375225709264, loss:0.7575111825241072
Epoch:5, w1:0.78

(1.856752671962907,
 -1.1001763730425846,
 0.5575426233221273,
 3.928685687479997,
 1.8867592746797475,
 1.9297668859604287,
 -3.6206378734575115)

keras gave us w1=1.8236613, w2=-1.599353, w3=0.06977581, w4 = 6.785558, w5=2.07437, w6=2.6513433, bias =-4.982498

and our custom gradient descent is giving us w1= 1.856752671962907,
w2 = -1.1001763730425846,
w3 = 0.5575426233221273,
w4 = 3.928685687479997,
w5 = 1.8867592746797475,
w6 = 1.9297668859604287,
bias = -3.6206378734575115. Almost same parameters

**Now we will create a class NN which has this gradient descent function which we will use to train our model**

In [51]:
class myNeuralNetwork():
  def __init__(self):
    self.w1 = 1,
    self.w2 = 1,
    self.w3 = 1,
    self.w4 = 1,
    self.w5 = 1,
    self.w6 = 1,
    self.bias = 0

  def fit(self, X, y, epochs, loss_threshold):
    self.w1, self.w2, self.w3, self.w4, self.w5, self.w6, self.bias = self.gradient_descent(X['Pregnancies'],X['BloodPressure'],X['SkinThickness'],X['BMI'],X['DiabetesPedigreeFunction'],X['Age'], y, epochs, loss_threshold)

  def predict(self, X_test):
    weighted_sum =  self.w1 * X_test['Pregnancies'] + self.w2 * X_test['BloodPressure'] + self.w3 * X_test['SkinThickness'] + self.w4 * X_test['BMI']+ self.w5 * X_test['DiabetesPedigreeFunction']+ self.w6 * X_test['Age'] + self.bias
    return sigmoid_numpy(weighted_sum)

  def gradient_descent(self, Pregnancies, BloodPressure, SkinThickness, BMI,
       DiabetesPedigreeFunction, Age, y_true, epochs, loss_thresold):
    w1 = w2 = w3 = w4 = w5 = w6 = 1
    bias = 0
    rate = 0.5
    n = len(Pregnancies)
    for i in range(epochs):
        weighted_sum = w1 * Pregnancies + w2 * BloodPressure + w3 * SkinThickness + w4 * BMI + w5 * DiabetesPedigreeFunction + w6 * Age +bias
        y_predicted = sigmoid_numpy(weighted_sum)
        loss = log_loss(y_true, y_predicted)

        w1d = (1/n)*np.dot(np.transpose(Pregnancies),(y_predicted-y_true))
        w2d = (1/n)*np.dot(np.transpose(BloodPressure),(y_predicted-y_true))
        w3d = (1/n)*np.dot(np.transpose(SkinThickness),(y_predicted-y_true))
        w4d = (1/n)*np.dot(np.transpose(BMI),(y_predicted-y_true))
        w5d = (1/n)*np.dot(np.transpose(DiabetesPedigreeFunction),(y_predicted-y_true))
        w6d = (1/n)*np.dot(np.transpose(Age),(y_predicted-y_true))

        bias_d = np.mean(y_predicted - y_true)
        w1 = w1 - rate * w1d
        w2 = w2 - rate * w2d
        w3 = w3 - rate * w3d
        w4 = w4 - rate * w4d
        w5 = w5 - rate * w5d
        w6 = w6 - rate * w6d
        bias = bias - rate * bias_d

        if i%50 == 0:
          print (f'Epoch:{i}, w1:{w1}, w2:{w2}, w3:{w3}, w4:{w4}, w5:{w5}, w6:{w6}, bias:{bias}, loss:{loss}')

        if loss<=loss_thresold:
          print (f'Epoch:{i}, w1:{w1}, w2:{w2}, w3:{w3}, w4:{w4}, w5:{w5}, w6:{w6}, bias:{bias}, loss:{loss}')
          break

    return w1, w2, w3, w4, w5, w6, bias

In [52]:
X_train_scaled[:5]

Pregnancies  BloodPressure  SkinThickness       BMI  \
118     0.235294       0.491803       0.232323  0.420268   
336     0.000000       0.000000       0.000000  0.503726   
374     0.117647       0.426230       0.434343  0.539493   
518     0.764706       0.491803       0.000000  0.488823   
716     0.176471       0.639344       0.393939  0.503726   

     DiabetesPedigreeFunction       Age  
118                  0.183058  0.271605  
336                  0.385124  0.543210  
374                  0.337190  0.345679  
518                  0.074380  0.506173  
716                  0.400826  0.382716

In [54]:
mycustomModel = myNeuralNetwork()
mycustomModel.fit(X_train_scaled, y_train, epochs =3500 , loss_threshold= 0.5479)

Epoch:0, w1:0.9484038037549335, w2:0.8482495061987809, w3:0.9464823618189773, w4:0.8811597148577026, w5:0.9526961436461479, w6:0.8966941323397016, bias:-0.26403054991935676, loss:1.4049969349785973
Epoch:50, w1:0.9397659693535919, w2:0.03847574991676495, w3:0.7035572250090617, w4:0.5703953573687155, w5:0.8529990151243797, w6:0.585024889287134, bias:-1.644255299281763, loss:0.6072835996360086
Epoch:100, w1:1.1432812047474523, w2:-0.0808582952798168, w3:0.7101752792697007, w4:0.8152990697693228, w5:0.9724377849653247, w6:0.7367869146719543, bias:-1.8349215986695957, loss:0.5996692308648538
Epoch:150, w1:1.3012711859711785, w2:-0.18858571459173182, w3:0.7159053328882602, w4:1.0521716488405626, w5:1.0806640829673628, w6:0.8676329602181588, bias:-2.006380388372145, loss:0.5936039063975715
Epoch:200, w1:1.4243202967951334, w2:-0.2861470858984741, w3:0.7199491111120895, w4:1.2805929511648542, w5:1.178413498416561, w6:0.9820254083244705, bias:-2.1618360356086086, loss:0.5886381425279097
Epoch:

we are getting values as follows:Epoch:3450, w1:1.827081486018313, w2:-1.5334488163802882, w3:0.10901974537797972, w4:6.609883112323265, w5:2.065341944940846, w6:2.646975292401594, bias:-4.947070630825059, loss:0.5479874280764988

In [57]:
mycustomModel.predict(X_test_scaled)[:5]

459    0.671991
39     0.786205
344    0.713529
84     0.610109
700    0.297448
dtype: float64

In [58]:
model.predict(X_test_scaled)[:5]

5/5 [==============================] - 0s 2ms/step


array([[0.66839594],
       [0.78786856],
       [0.71803695],
       [0.6174387 ],
       [0.2994859 ]], dtype=float32)

So we can see from above that both keras neural network and our custom made neural network for diabetes classification are giving approximately same results.